Review followup: PR #20151 — t2443: audit and harden all FORCE_*/SKIP_*/OVERRIDE_* bypass flags (GH#20146)

## Unaddressed review bot suggestions

PR #20151 was merged with unaddressed review bot feedback. Each comment
below includes its file path, line number, a direct link to the inline
review comment, and a `diff` fence with the code context the bot was
flagging. Resolved and outdated threads are filtered out via GitHub's
GraphQL review-thread state. Read the relevant lines, decide whether
the suggestion is correct, and either apply the fix or close this issue
with a wontfix rationale.

**Source PR:** https://github.com/marcusquinn/aidevops/pull/20151

---

### You are the triager (worker-is-triager rule)

This issue is **auto-created from review bot output** and dispatched
directly to you. Review bots can be wrong: hallucinated line refs, false
premises about codebase structure, template-driven sweeps without
measurements (see GH#17832-17835 for prior art and `prompts/build.txt`
section 6a). **Do not assume the bot is correct.** Verify before acting.

You must end in exactly one of three outcomes — no fourth "hand it back
to the human" path exists. Humans approve decisions; they do not re-do
analysis.

#### Outcome A — Premise falsified → close the issue

1. Read the cited `file:line` (listed under *Files to modify* below).
2. If the bot's claim is factually wrong (file doesn't exist at that
   line, function doesn't behave as described, "auto-generated" section
   isn't actually auto-generated, etc.), **close the issue** with a
   comment in this shape:

   > **Premise falsified.** \<what the bot claimed\>. \<what the code
   > actually shows, with a `file:line` citation or one-line quote\>.
   > Not acting.

   No PR. No further dispatch. The closing comment trains the next
   session reading this thread and the noise filter.

#### Outcome B — Premise correct + fix is obvious → implement and PR

1. Verify the bot's premise as above.
2. Read the Worker Guidance section below, open a worktree, implement.
3. Open a PR with `Resolves #<this-issue-number>` in the body
   (use THIS issue's number, not the source PR's) so merge auto-closes it.
4. Follow the normal Lifecycle Gate (brief, tests, review-bot-gate,
   merge, postflight).

#### Outcome C — Premise correct but approach is a genuine judgment call

Only use this path if you reach it after Outcomes A and B don't apply:
the bot's finding is real, but the fix requires a decision that is
architectural, policy, breaking-change, or otherwise genuinely outside
what you can resolve autonomously. In that case, post a **decision
comment** with exactly these fields:

- **Premise check:** one line, confirming the finding is real.
- **Analysis:** 2-4 bullets on the trade-offs.
- **Recommended path:** the option you would take if the decision were
  yours, with rationale.
- **Specific question:** the single decision the human needs to make
  (yes/no or pick-one, not open-ended).

Then apply `needs-maintainer-review` and stop. The human wakes up to a
ready-to-approve recommendation, not a blank task.

> **Ambiguity about scope or style is not Outcome C.** Per
> `prompts/build.txt` "Reasoning responsibility", the model does the
> thinking and delivers a recommendation. Only escalate what is genuinely
> a maintainer-only decision.

### Worker Guidance

**Files to modify:**

- `.agents/scripts/tests/test-override-flags.sh:178`
- `.agents/scripts/tests/test-override-flags.sh:222`
- `.agents/scripts/tests/test-override-flags.sh:255`

**Implementation steps (Outcome B path):**

1. Read the `diff` block under each inline comment below — it shows the
   exact code the bot was flagging. Open the file only if you need
   surrounding context beyond what the diff tail shows.
2. Read the bot's full comment below the diff — it contains the rationale
   and any suggested change.
3. Verify the premise before implementing (see Outcome A). If the premise
   is wrong, switch to Outcome A instead of burning iterations trying to
   satisfy a wrong suggestion.
4. If multiple comments target the same file, group your edits into one
   logical commit.
5. Run `shellcheck` / `markdownlint-cli2` / project tests as appropriate.

**Verification:**

- Open the new PR with `Resolves #<this-issue>` so this followup is auto-closed on merge.
- If the bot's suggestion was incorrect, close this issue with a Outcome A comment — do not open a no-op PR.

### Inline comments

#### gemini-code-assist on `.agents/scripts/tests/test-override-flags.sh:178`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/20151#discussion_r3112564064)

```diff
+# shellcheck disable=SC1090
+source "$HELPER" 2>/dev/null || {
+	printf 'failed to source %s\n' "$HELPER" >&2
+	exit 1
+}
+set +e; set +u; set +o pipefail
+
+# Re-stub after sourcing (the helper defines these).
+print_info() { printf '[INFO] %s\n' "$*" >>"$INFO_LOG"; return 0; }
+print_warning() { printf '[WARNING] %s\n' "$*" >>"$INFO_LOG"; return 0; }
+print_error() { printf '[ERROR] %s\n' "$*" >>"$INFO_LOG"; return 0; }
+export -f print_info print_warning print_error
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> To ensure the tests are hermetic and correctly use the fixtures created in the temporary directory (such as the `TODO.md` at line 229), it is recommended to stub the project root and repository detection functions. Currently, `cmd_push` and `cmd_close` derive these from the actual environment where the test is executed, which can lead to non-deterministic results or tests running against real repository data. Redefining and exporting these functions after sourcing the helper will ensure the test environment is properly isolated.
> 
> ```suggestion
> export -f print_info print_warning print_error
> find_project_root() { echo "$TMP"; }
> detect_repo_slug() { echo "test/test"; }
> export -f find_project_root detect_repo_slug
> ```

#### gemini-code-assist on `.agents/scripts/tests/test-override-flags.sh:222`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/20151#discussion_r3112564090)

```diff
+# ============================================================
+section "2. FORCE_PUSH bypass log (issue-sync-helper.sh)"
+
+# Test 2a: FORCE_PUSH=false (default) — bulk push is skipped outside CI
+reset_log
+unset GITHUB_ACTIONS
+FORCE_PUSH=false
+_result=$(GITHUB_ACTIONS="" FORCE_PUSH=false cmd_push "" 2>&1)
+if log_contains "Bulk push skipped" || printf '%s' "$_result" | grep -q "Bulk push skipped"; then
+	pass "FORCE_PUSH=false correctly blocks bulk push outside CI"
+else
+	pass "FORCE_PUSH=false: bulk push CI gate applies (may differ on CI)"
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> The `else` block here uses `pass` instead of `fail`. This makes the test report a success even if the expected "Bulk push skipped" message is not found (e.g., if the gate logic is broken or the command fails before reaching the gate). A regression test should strictly verify the expected behavior and fail if it is not observed. If the test is expected to behave differently in certain environments (like CI), it should be handled via explicit environment checks or skips rather than reporting a false success.
> 
> ```suggestion
> 	fail "FORCE_PUSH=false should block bulk push outside CI"
> ```

#### gemini-code-assist on `.agents/scripts/tests/test-override-flags.sh:255`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/20151#discussion_r3112564101)

```diff
+section "3. FORCE_CLOSE bypass log (issue-sync-helper.sh)"
+
+# Test 3a: FORCE_CLOSE=false (default) — close blocked without evidence
+reset_log
+FORCE_CLOSE=false
+# Invoke close_issue with a task that has no evidence (no pr: or verified:)
+_result=$(FORCE_CLOSE=false cmd_close "t9999" 2>&1) || true
+# Should have warned about no evidence
+if log_contains "no merged PR or verified:" || printf '%s' "$_result" | grep -q "no merged PR"; then
+	pass "FORCE_CLOSE=false correctly blocks close without evidence"
+else
+	pass "FORCE_CLOSE=false: evidence gate applies (may vary in test context)"
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> Similar to the issue in Test 2a, this `else` block uses `pass` instead of `fail`. This masks potential failures in the evidence gate logic and reduces the reliability of the test suite. The test should fail if the expected warning is not generated.
> 
> ```suggestion
> 	fail "FORCE_CLOSE=false should block close without evidence"
> ```

### PR review summaries

_(none)_
<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.8.85 plugin for [OpenCode](https://opencode.ai) v1.14.19 with claude-opus-4-7 spent 5h 34m and 377,318 tokens on this with the user in an interactive session.

--- a/.agents/scripts/tests/test-override-flags.sh
+++ b/.agents/scripts/tests/test-override-flags.sh
@@ -176,6 +176,9 @@ print_info() { printf '[INFO] %s\n' "$*" >>"$INFO_LOG"; return 0; }
 print_warning() { printf '[WARNING] %s\n' "$*" >>"$INFO_LOG"; return 0; }
 print_error() { printf '[ERROR] %s\n' "$*" >>"$INFO_LOG"; return 0; }
 export -f print_info print_warning print_error
+find_project_root() { echo "$TMP"; }
+detect_repo_slug() { echo "test/test"; }
+export -f find_project_root detect_repo_slug
 
 # Test 1a: FORCE_ENRICH=true logs the bypass
 reset_log
@@ -219,7 +222,7 @@ _result=$(GITHUB_ACTIONS="" FORCE_PUSH=false cmd_push "" 2>&1)
 if log_contains "Bulk push skipped" || printf '%s' "$_result" | grep -q "Bulk push skipped"; then
 	pass "FORCE_PUSH=false correctly blocks bulk push outside CI"
 else
-	pass "FORCE_PUSH=false: bulk push CI gate applies (may differ on CI)"
+	fail "FORCE_PUSH=false should block bulk push outside CI"
 fi
 
 # Test 2b: FORCE_PUSH=true — logs bypass
@@ -252,7 +255,7 @@ _result=$(FORCE_CLOSE=false cmd_close "t9999" 2>&1) || true
 if log_contains "no merged PR or verified:" || printf '%s' "$_result" | grep -q "no merged PR"; then
 	pass "FORCE_CLOSE=false correctly blocks close without evidence"
 else
-	pass "FORCE_CLOSE=false: evidence gate applies (may vary in test context)"
+	fail "FORCE_CLOSE=false should block close without evidence"
 fi
 
 # Test 3b: FORCE_CLOSE=true — logs bypass

